File: lib/languages/go/function-definitions.go
package main

import (
    "fmt"
    "strings"
)

// Basic function definition
func greet(name string) string {
    return "Hello, " + name
}

// Function with multiple parameters
func add(a, b int) int {
    return a + b
}

// Function with multiple return values
func divide(a, b float64) (float64, error) {
    if b == 0 {
        return 0, fmt.Errorf("division by zero")
    }
    return a / b, nil
}

// Function with named return values
func calculateStats(values []int) (min, max, sum int) {
    if len(values) == 0 {
        return 0, 0, 0
    }

    min = values[0]
    max = values[0]
    sum = 0

    for _, v := range values {
        if v < min {
            min = v
        }
        if v > max {
            max = v
        }
        sum += v
    }

    return // "naked" return - returns the named values
}

// Variadic function (variable number of arguments)
func sum(numbers ...int) int {
    total := 0
    for _, num := range numbers {
        total += num
    }
    return total
}

// Method definition (function with a receiver)
type Person struct {
    FirstName string
    LastName  string
}

// Value receiver
func (p Person) FullName() string {
    return p.FirstName + " " + p.LastName
}

// Pointer receiver
func (p *Person) SetName(firstName, lastName string) {
    p.FirstName = firstName
    p.LastName = lastName
}

func main() {
    // Calling basic functions
    message := greet("Gopher")
    fmt.Println(message)

    result := add(5, 3)
    fmt.Println("Sum:", result)

    // Handling multiple return values
    quotient, err := divide(10, 2)
    if err != nil {
        fmt.Println("Error:", err)
    } else {
        fmt.Println("Quotient:", quotient)
    }

    // Using named return values
    numbers := []int{7, 2, 9, 3, 5}
    min, max, total := calculateStats(numbers)
    fmt.Println("Min:", min, "Max:", max, "Sum:", total)

    // Calling variadic function
    fmt.Println("Sum:", sum(1, 2, 3, 4, 5))

    // Spreading a slice to a variadic function
    nums := []int{1, 2, 3, 4, 5}
    fmt.Println("Sum from slice:", sum(nums...))

    // Using methods
    person := Person{FirstName: "John", LastName: "Doe"}
    fmt.Println("Full name:", person.FullName())

    person.SetName("Jane", "Smith")
    fmt.Println("New full name:", person.FullName())
}

File: lib/languages/zig/function-definitions.zig
// Basic function with type annotations
fn add(a: i32, b: i32) i32 {
    return a + b;  // Explicit return
}

// Function with multiple returns using error union
fn divide(a: f64, b: f64) !f64 {
    if (b == 0.0) {
        return error.DivisionByZero;
    }
    return a / b;
}

// Function with optional return
fn findItem(items: []const i32, value: i32) ?usize {
    for (items, 0..) |item, i| {
        if (item == value) {
            return i;
        }
    }
    return null;
}

// Generic function using comptime
fn min(comptime T: type, a: T, b: T) T {
    return if (a < b) a else b;
}

// Function with multiple return values using tuple
fn divMod(a: i32, b: i32) struct { quotient: i32, remainder: i32 } {
    return .{
        .quotient = a / b,
        .remainder = a % b,
    };
}

// Function with variadic arguments
fn sum(args: anytype) i32 {
    var result: i32 = 0;
    inline for (args) |arg| {
        result += arg;
    }
    return result;
}

// Calling variadic function
const total = sum(.{1, 2, 3, 4});  // 10

// Function with pointers
fn increment(value: *i32) void {
    value.* += 1;
}

// Calling function with pointer
var x: i32 = 5;
increment(&x);  // x is now 6

